[Jobs] Make sure JobConfig specified in the init script is not being overridden by the SupervisorActor

alexeykudinkin opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

NOTE: This is a revert of the [revert](https://github.com/ray-project/ray/pull/38655), addressing broken tests

Currently when executing the job submitted via Ray Job submission API, `SupervisorActor` will be overriding whole of the `JobConfig` making it impossible to specify additional params like `JobConfig.code_search_path` for ex.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

alexeykudinkin commented on python/ray/_private/worker.py:
```diff
@@ -1416,16 +1419,26 @@ def init(
         if ray_constants.RAY_RUNTIME_ENV_HOOK in os.environ and not _skip_env_hook:
             runtime_env = _load_class(os.environ[ray_constants.RAY_RUNTIME_ENV_HOOK])(
-                job_config.runtime_env
+                runtime_env
```
This line is the only change as compared to https://github.com/ray-project/ray/pull/38655

architkulkarni commented:
@alexeykudinkin potentially related test failure  https://buildkite.com/ray-project/oss-ci-build-pr/builds/32913#018a198f-5512-49d0-9b6b-f4d11026424e/3356-8245 (or https://buildkite.com/ray-project/oss-ci-build-pr/builds/32913#018a198f-5512-49d0-9b6b-f4d11026424e/3356-8229 or https://buildkite.com/ray-project/oss-ci-build-pr/builds/32913#018a198f-5512-49d0-9b6b-f4d11026424e/3356-8218, not sure which one is the root cause)

edoakes commented:
Test failure: https://buildkite.com/ray-project/oss-ci-build-pr/builds/32913#018a198f-5512-49d0-9b6b-f4d11026424e/3356-8768

alexeykudinkin commented:
Thanks for reviewing @edoakes @architkulkarni!

Yeah, i figured that new test that i've added was messing with our Ray cluster sharing in other tests and fixed that, but when i've tried to run other tests locally it turned out that there quite a few more tests that were failing in `test_job_manager`: basically all of the tests doing `ray start --head` seem to have been broken for quite some time (b/c these are not provided `--address` flag which seemed to be expected by `call_ray_start` fixture (and it's broken on master as well).

Regardless, i've fixed that one as well, and now it should be working fine.

architkulkarni commented on python/ray/tests/conftest.py:
```diff
@@ -589,11 +589,16 @@ def call_ray_start_context(request):
-    address_location = out.find(redis_substring_prefix) + len(redis_substring_prefix)
-    address = out[address_location:]
-    address = address.split("'")[0]
+    idx = out.find(redis_substring_prefix)
```
Ah I think the discussion here might be related https://github.com/ray-project/ray/pull/35226. It got reverted for breaking other tests, sorry for not following through and unreverting it.

alexeykudinkin commented:
Checked out test failures and all seems unrelated.

@edoakes @architkulkarni can you folks help merging this in?

architkulkarni commented:
@alexeykudinkin https://buildkite.com/ray-project/premerge/builds/2812#018a1e80-6f06-4fa9-8fd0-9a08f35221ca/166-406 related?

alexeykudinkin commented:
@architkulkarni fixed linter's failures.

Current failures don't seem to be related.

architkulkarni commented:
Failures
- Windows tests:test_streaming_generator::test_status_task_events_metrics is not flaky on master. But it seems unrelated
- rllib:TestLearnerGroupAsyncUpdate flaky on master and unrelated
- tune:test_commands flaky on master and unrelated
- chaos network delay test flaky on master and unrelated
